### ISSUE TITLE
some random thoughts

### ISSUE THREAD

**Issue by cgwalters** [travier/ostree-containers#3]
One thing I *definitely* like is being able to define containers in a declarative, structured way that can be managed in a git-ops fashion.  That's how rpm-ostree was always designed.

However that said, I do also think we should try hard not to create a divergence between "base images" and "application images" and layered images.

In a nutshell, even though the nginx image (and toolbox) in fedora is a base image, that's just for bad reasons because there's not a container-native buildsystem deployed there.  Wouldn't you agree it'd make more sense for an image like that to *logically* derive from a base?

One problem with rpm-ostree for `container: true` today is that while it technically works, some "ostree stuff" leaks into the container images that we probably don't want in the future.

The work we did on "content addressable reproducible chunking" in the ostree stack is something I'd really like to support generalizing to `Containerfile` based builds too.  It's tricky though.

I also quite like rpm-ostree's change detection in builds.

But on the definite negatives, the rpm-ostree flow doesn't feel container native in other ways (injecting container metadata is a big one).

So I guess...medium/long term I think it's going to be more fruitful to take these good ideas from rpm-ostree and put them somewhere else.

For example, I'd really love to see `dnf` learn to parse *a* declarative format by default (whether that's toml or yaml or kickstart or something else I don't care too much to start).  Maybe it's even a spec file with just `Requires`!  And the declarative bits should be droppable in `/usr/lib/dnf/configs.d` and `/etc/dnf/configs.d`.

**Comment by travier** [comment]
> However that said, I do also think we should try hard not to create a divergence between "base images" and "application images" and layered images.
> 
> In a nutshell, even though the nginx image (and toolbox) in fedora is a base image, that's just for bad reasons because there's not a container-native buildsystem deployed there. Wouldn't you agree it'd make more sense for an image like that to _logically_ derive from a base?

I think this really depends of the use cases and the more it goes, the more I think we should make building special purpose images easier. The equivalent buzzword for those images is "distroless".

The split between a base image and layers does not always make sense:

- For the toolbox image, the layer was (before the move to kickstarts) and would be significantly bigger than the base image, thus a lot of network bandwidth would be wasted pulling layer updates (which needs the work you mention below to split layers). The previous layered build using a Containerfile was complex because it needed to undo a lot of the special cleaning that we need to keep the base image small as it is targeted for applications, not interactive usage.

- For the nginx image, we don't need most of the content of the base image such as `dnf` & `python`, so the layer would be removals, thus wasting space, and it is generally more difficult to remove packages from an image than to not include them.

In both cases, building those images directly with `rpm-ostree compose image` results in better results.

We'll still need generic base images for a while though so they are definitely here to stay.

Overall, I think there is a place for making base images leaner, while still letting users derive them with a Containerfile.

Base images would be were the package selection and installation happens and then other changes such as external packages, binaries, configuration changes can be be added via a Containerfile.

This is even useful for multi-staged builds as you get a truly minimal image without `dnf` at the end and only the packages that you need as runtime dependencies for your application.

> One problem with rpm-ostree for `container: true` today is that while it technically works, some "ostree stuff" leaks into the container images that we probably don't want in the future.

Yes, there are a few duplicated files (mostly the rpmdb) that we would have to remove to match 1:1 with the current images but that's mostly it from what I could find.

> The work we did on "content addressable reproducible chunking" in the ostree stack is something I'd really like to support generalizing to `Containerfile` based builds too. It's tricky though.

Yes, we would need something like that to make base images interesting for large layers.

> I also quite like rpm-ostree's change detection in builds.

Agree. I'm not using it right now because I'm blocked by https://github.com/coreos/rpm-ostree/issues/4782.

**Comment by cgwalters** [comment]
> I think this really depends of the use cases and the more it goes, the more I think we should make building special purpose images easier. The equivalent buzzword for those images is "distroless".

Fair point, but...mmm...I am not so sure about how hard to push this really.
